### 1.2 Propose an alternative encoding for the following visualization. Compare the visualizations based on perception. (10 points)
Either hand-draw or use an application to create a sketched solution. Upload an image and describe the differences between your solution and the FiveThirtyEight image in terms of perception (specifically for the task of comparing one movie to another).
!["Drowing"](assets/have_seen_resized.png)

Provide your answer here:

!["Answer12"](assets/sample.png)

There are several diff erences between these two visualizations. The most striking one is the mark type used. The 538 visualization is a horizontal bargraph comparing percentages, while the alternative visualization I created is a radar graph that illustrates the absolute number of respondents whohad viewed each movie, as opposed to showing the strict percentage. Therefore, the percentage of survey respondents who had viewed each movieis moreso implied than explicit.
<br><br>
In terms of the task of comparing one movie to another, my alternative visualization certainly illustrates that, but it may require a slightly higher eff ortto determine the exact numerical diff erences in viewership for each movie. The general trend is evident, but it is more diffi cult to extract the raw datathan in the original visualization. However, my visualization is perhaps more interesting or engaging to look at.

### 1.3 Propose an alternative encoding for the following visualization. Compare the visualizations based on perception. (10 points)
Again, either-hand draw or use an application to create a sketched solution. Upload an image and describe the differences between your solution and the FiveThirtyEight image in terms of perception (specifically for the task of comparing one movie to another).
!["Drowing"](assets/how_rate_resized.png)

In [79]:
import altair as alt
import pandas as pd

data_movies = pd.DataFrame()

# For each film...
data_movies.at['The Phantom Menace', 'Top Third'] = .16
data_movies.at['The Phantom Menace', 'Middle Third'] = .375
data_movies.at['The Phantom Menace', 'Bottom Third'] = .465
data_movies.at['Attack of the Clones', 'Top Third'] = .14
data_movies.at['Attack of the Clones', 'Middle Third'] = .29
data_movies.at['Attack of the Clones', 'Bottom Third'] = .57
data_movies.at['Revenge of the Sith', 'Top Third'] = .13
data_movies.at['Revenge of the Sith', 'Middle Third'] = .40
data_movies.at['Revenge of the Sith', 'Bottom Third'] = .47
data_movies.at['A New Hope', 'Top Third'] = .5
data_movies.at['A New Hope', 'Middle Third'] = .31
data_movies.at['A New Hope', 'Bottom Third'] = .19
data_movies.at['The Empire Strikes Back', 'Top Third'] = .64
data_movies.at['The Empire Strikes Back', 'Middle Third'] = .22
data_movies.at['The Empire Strikes Back', 'Bottom Third'] = .14
data_movies.at['Return of the Jedi', 'Top Third'] = .425
data_movies.at['Return of the Jedi', 'Middle Third'] = .405
data_movies.at['Return of the Jedi', 'Bottom Third'] = .17

data_movies2 = pd.DataFrame()

for col in data_movies.columns:
    for i in data_movies.index:
        row = [i, col, data_movies.loc[i][col]]
        data_movies2 = pd.concat([data_movies2, pd.DataFrame(row).T.rename(columns = {0: 'Movie', 1: 'type', 2: 'value'})])

data_movies2.set_index('Movie', inplace = True)

episodes = ['EI', 'EII', 'EIII', 'EIV', 'EV', 'EVI']
names = {
    'EI' : 'The Phantom Meanance', 'EII' : 'Attack of the clones', 'EIII' : 'Revenge of the Sith', 
    'EIV': 'A New Hope', 'EV': 'The Empire Strikes Back', 'EVI' : 'The Return of the Jedi'
}

# we're also going to use this order to sort, so names_l will now have our sort order
names_l = [names[ep] for ep in episodes]

bars = alt.Chart(data_movies2.reset_index()).mark_bar().encode(
    x=alt.X('value:Q', axis=None),
    y=alt.Y('Movie', sort=names_l, title=""),
    color='type:N'
    
).properties(
    height = 300,
    width = 600
)


bars

alt.Chart(...)

## Part 2. Altair programming exercise (70 points)
We have provided you with some code and parts of the article [America’s Favorite ‘Star Wars’ Movies (And Least Favorite Characters)](https://fivethirtyeight.com/features/americas-favorite-star-wars-movies-and-least-favorite-characters/). This article is based on the dataset:

1. [StarWars](assets/StarWars.csv) Created by FiveThirtyEight based on a survey ran through SurveyMonkey Audience, surveying 1,186 respondents from June 3 to 6 2014. Available [online] (https://github.com/fivethirtyeight/data/tree/master/star-wars-survey)

To earn points for this assignment, you must:

- Recreate the visualizations in the article (replace the images in the article with a code cell that creates a visualization). We provide one example. Each visualization is worth 10 points (40 points/ 10 each x 4 total ).

    - _Partial credit can be granted for each visualization (up to 5 points) if you provide the grammar of graphics description of the visualization without a functional Altair implementation_


- Propose one alternative visualization for one of the article visualizations. Add a short paragraph describing why your visualization is more *effective* based on principles of perception/cognition. (15 points/ 10 points plot + 5 justification)


- Propose a new visualization to complement a part of the article. Add a short paragraph justifying your decisions in terms of Perception/Cognition processes.  (15 points/ 10 points plot + 5 justification)


In [24]:
import pandas as pd
import altair as alt
import numpy as np
import math

pd.set_option("display.max_columns", 50)

In [25]:
# enable correct rendering
alt.renderers.enable('default')

RendererRegistry.enable('default')

In [26]:
# uses intermediate json files to speed things up
alt.data_transformers.enable('json')

DataTransformerRegistry.enable('json')

In [27]:
sw = pd.read_csv('assets/StarWars.csv', encoding='latin1')

# Some format is needed for the survey dataframe, we provide the formatted dataset in a dataframe 
sw = sw.rename(columns={'Have you seen any of the 6 films in the Star Wars franchise?':'seen_any_movie',
                        'Do you consider yourself to be a fan of the Star Wars film franchise?': 'fan',
                        'Which of the following Star Wars films have you seen? Please select all that apply.' : 'seen_EI',
                        'Unnamed: 4' : 'seen_EII',
                        'Unnamed: 5' : 'seen_EIII',
                        'Unnamed: 6' : 'seen_EIV',
                        'Unnamed: 7' : 'seen_EV',
                        'Unnamed: 8' : 'seen_EVI',
                        'Please rank the Star Wars films in order of preference with 1 being your favorite film in the franchise and 6 being your least favorite film.' : 'rank_EI',
                        'Unnamed: 10' : 'rank_EII',
                        'Unnamed: 11' : 'rank_EIII',
                        'Unnamed: 12' : 'rank_EIV',
                        'Unnamed: 13' : 'rank_EV',
                        'Unnamed: 14' : 'rank_EVI',
                        'Please state whether you view the following characters favorably, unfavorably, or are unfamiliar with him/her.' : 'Han Solo',
                        'Unnamed: 16' : 'Luke Skywalker',
                        'Unnamed: 17' : 'Princess Leia Organa',
                        'Unnamed: 18' : 'Anakin Skywalker',
                        'Unnamed: 19' : 'Obi Wan Kenobi',
                        'Unnamed: 20' : 'Emperor Palpatine',
                        'Unnamed: 21' : 'Darth Vader',
                        'Unnamed: 22' : 'Lando Calrissian',
                        'Unnamed: 23' : 'Boba Fett',
                        'Unnamed: 24' : 'C-3P0',
                        'Unnamed: 25' : 'R2 D2',
                        'Unnamed: 26' : 'Jar Jar Binks',
                        'Unnamed: 27' : 'Padme Amidala',
                        'Unnamed: 28' : 'Yoda',
                       })
sw = sw.drop([0])

In [28]:
# take a peak to look at the data
sw.head(5)

RespondentID seen_any_movie  fan                                   seen_EI  \
1  3.292880e+09            Yes  Yes  Star Wars: Episode I  The Phantom Menace   
2  3.292880e+09             No  NaN                                       NaN   
3  3.292765e+09            Yes   No  Star Wars: Episode I  The Phantom Menace   
4  3.292763e+09            Yes  Yes  Star Wars: Episode I  The Phantom Menace   
5  3.292731e+09            Yes  Yes  Star Wars: Episode I  The Phantom Menace   

                                      seen_EII  \
1  Star Wars: Episode II  Attack of the Clones   
2                                          NaN   
3  Star Wars: Episode II  Attack of the Clones   
4  Star Wars: Episode II  Attack of the Clones   
5  Star Wars: Episode II  Attack of the Clones   

                                     seen_EIII  \
1  Star Wars: Episode III  Revenge of the Sith   
2                                          NaN   
3  Star Wars: Episode III  Revenge of the Sith   
4  Star Wars: Episode III  Revenge of the Sith   
5  Star Wars: Episode III  Revenge of the Sith   

                            seen_EIV  \
1  Star Wars: Episode IV  A New Hope   
2                                NaN   
3                                NaN   
4  Star Wars: Episode IV  A New Hope   
5  Star Wars: Episode IV  A New Hope   

                                        seen_EV  \
1  Star Wars: Episode V The Empire Strikes Back   
2                                           NaN   
3                                           NaN   
4  Star Wars: Episode V The Empire Strikes Back   
5  Star Wars: Episode V The Empire Strikes Back   

                                   seen_EVI rank_EI rank_EII rank_EIII  \
1  Star Wars: Episode VI Return of the Jedi       3        2         1   
2                                       NaN     NaN      NaN       NaN   
3                                       NaN       1        2         3   
4  Star Wars: Episode VI Return of the Jedi       5        6         1   
5  Star Wars: Episode VI Return of the Jedi       5        4         6   

  rank_EIV rank_EV rank_EVI            Han Solo      Luke Skywalker  \
1        4       5        6      Very favorably      Very favorably   
2      NaN     NaN      NaN                 NaN                 NaN   
3        4       5        6  Somewhat favorably  Somewhat favorably   
4        2       4        3      Very favorably      Very favorably   
5        2       1        3      Very favorably  Somewhat favorably   

  Princess Leia Organa      Anakin Skywalker      Obi Wan Kenobi  \
1       Very favorably        Very favorably      Very favorably   
2                  NaN                   NaN                 NaN   
3   Somewhat favorably    Somewhat favorably  Somewhat favorably   
4       Very favorably        Very favorably      Very favorably   
5   Somewhat favorably  Somewhat unfavorably      Very favorably   

    Emperor Palpatine         Darth Vader  \
1      Very favorably      Very favorably   
2                 NaN                 NaN   
3    Unfamiliar (N/A)    Unfamiliar (N/A)   
4  Somewhat favorably      Very favorably   
5    Very unfavorably  Somewhat favorably   

                              Lando Calrissian             Boba Fett  \
1                             Unfamiliar (N/A)      Unfamiliar (N/A)   
2                                          NaN                   NaN   
3                             Unfamiliar (N/A)      Unfamiliar (N/A)   
4                           Somewhat favorably  Somewhat unfavorably   
5  Neither favorably nor unfavorably (neutral)        Very favorably   

                C-3P0               R2 D2     Jar Jar Binks  \
1      Very favorably      Very favorably    Very favorably   
2                 NaN                 NaN               NaN   
3    Unfamiliar (N/A)    Unfamiliar (N/A)  Unfamiliar (N/A)   
4      Very favorably      Very favorably    Very favorably   
5  Somewhat favorably  Somewhat favorably  Very unfavorably   

  

# America’s Favorite ‘Star Wars’ Movies (And Least Favorite Characters)

_Original article available at [FiveThirtyEight](https://fivethirtyeight.com/features/americas-favorite-star-wars-movies-and-least-favorite-characters/)_

By [Walt Hickey](https://fivethirtyeight.com/contributors/walt-hickey/)

Filed under [Movies](https://fivethirtyeight.com/tag/movies/)

Get the data on [GitHub](https://github.com/fivethirtyeight/data/tree/master/star-wars-survey)

This week, I caught a sneak peek [of the X-Wing fighter](http://www.wired.com/2014/07/star-wars-episode-vii-x-wing/) from the new “Star Wars” films in production. The forthcoming movies — and the middling response to the most recent trilogy — provide a perfect excuse to examine some questions I’ve long wanted answers to: How many people are “Star Wars” fans? Does the rest of America realize that “The Empire Strikes Back” is clearly the best of the bunch? Which characters are most well-liked and most hated? And who shot first, Han Solo or Greedo?

We ran a poll through [SurveyMonkey Audience](https://www.surveymonkey.com/mp/audience/), surveying 1,186 respondents from June 3 to 6 (the [data](https://github.com/fivethirtyeight/data/tree/master/star-wars-survey) is available [on GitHub](https://github.com/fivethirtyeight/data)). Seventy-nine percent of those respondents said they had watched at least one of the “Star Wars” films. This question, incidentally, had a substantial difference by gender: 85 percent of men have seen at least one “Star Wars” film compared to 72 percent of women. Of people who have seen a film, men were also more likely to consider themselves a fan of the franchise: 72 percent of men compared to 60 percent of women.

We then asked respondents which of the films they had seen. With 835 people responding, here’s the probability that someone has seen a given “Star Wars” film given that they have seen any Star Wars film:

!["Sol1"](assets/have_seen_resized.png)

In [29]:
# Sample visualization

# We're going to fix the labels a bit so will create a mapping to the full names
episodes = ['EI', 'EII', 'EIII', 'EIV', 'EV', 'EVI']
names = {
    'EI' : 'The Phantom Meanance', 'EII' : 'Attack of the clones', 'EIII' : 'Revenge of the Sith', 
    'EIV': 'A New Hope', 'EV': 'The Empire Strikes Back', 'EVI' : 'The Return of the Jedi'
}

# we're also going to use this order to sort, so names_l will now have our sort order
names_l = [names[ep] for ep in episodes]

print("sort order: ",names_l)

sort order:  ['The Phantom Meanance', 'Attack of the clones', 'Revenge of the Sith', 'A New Hope', 'The Empire Strikes Back', 'The Return of the Jedi']


In [30]:
# let's do some data pre-processing... sw (star wars) has everything

# We want to only use those people who have seen at least one movie, let's get the people, toss NAs
# and get the total count

# find people who have at least on of the columns (seen_*) not NaN
seen_at_least_one = sw.dropna(subset=['seen_' + ep for ep in episodes],how='all')
total = len(seen_at_least_one)

print("total who have seen at least one: ", total)

total who have seen at least one:  835


In [31]:
# for each movie, we're going to calculate the percents and generate a new data frame
percs = []

# loop over each column and calculate the number of people who have seen the movie
# specifically, filter out the people who are *NaN* for a specific episode (e.g., ep_EII), count them
# and divide by the percent
for seen_ep in ['seen_' + ep for ep in episodes]:
    perc = len(seen_at_least_one[~ pd.isna(seen_at_least_one[seen_ep])]) / total
    percs.append(perc)
    
# at this point percs is holding our percentages

# now we're going use a trick to make tuples--pairing names with percents--using "zip" and then make a dataframe
tuples = list(zip([names[ep] for ep in episodes],percs))
seen_per_df = pd.DataFrame(tuples, columns = ['Name', 'Percentage'])
seen_per_df

Name  Percentage
0     The Phantom Meanance    0.805988
1     Attack of the clones    0.683832
2      Revenge of the Sith    0.658683
3               A New Hope    0.726946
4  The Empire Strikes Back    0.907784
5   The Return of the Jedi    0.883832

In [32]:
# ok, time to make the chart... let's make a bar chart (use mark_bar)
bars = alt.Chart(seen_per_df).mark_bar(size=30).encode(
    # encode x as the percent, and hide the axis
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        # encode y using the name, use the movie name to label the axis, sort using the names_l
        'Name:N',
         axis=alt.Axis(tickCount=5, title='', labelFontSize=13,labelColor="grey", tickColor="grey", tickSize=10),
         # we give the sorting order to avoid alphabetical order
         sort=names_l
    )
)

# at this point we don't really have a great plot (it's missing the annotations, titles, etc.)
bars


alt.Chart(...)

In [33]:
# we're going to overlay the text with the percentages, so let's make another visualization
# that's just text labels

text = bars.mark_text(
    align='left',
    baseline='middle',
    dx=3,  # Nudges text to right so it doesn't appear on top of the bar
    #color="#008fd5"
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Percentage:Q',format='.0%'),
    
)

# finally, we're going to combine the bars and the text and do some styling
seen_movies = (text + bars).configure_mark(
    # we don't love the blue
    color='dodgerblue'
).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.1
).properties(
    # set the dimensions of the visualization
    width=600,
    height=300
).properties(
    # add a title
    title=alt.TitleParams("Which 'Star Wars' Movies Have you Seen?", subtitle=["Of 835 repondents who has seen any film"]
                         )
).configure_title(anchor="start", fontSize=25,offset=20, subtitleFontSize=14)


seen_movies

# note that we are NOT formatting this in the Five Thirty Eight Style yet... we'll leave that to you to figure out

alt.LayerChart(...)

So we can see that “Star Wars: Episode V — The Empire Strikes Back” is the film seen by the most number of people, followed by “Star Wars: Episode VI — Return of the Jedi.” Appallingly, more people reported seeing “Star Wars: Episode I — The Phantom Menace” than the original “Star Wars” (renamed “Star Wars: Episode IV — A New Hope”).

So, which movie is the best? We asked the subset of 471 respondents who indicated they have seen every “Star Wars” film to rank them from best to worst. From that question, we calculated the share of respondents who rated each film as their favorite.

!["Sol1"](assets/best_movie_article_resized.png)


** Homework note: Click [here](assets/best_movie.png) to see a version of this plot generated in Altair.

### 2.1 What's the best 'Star Wars' movie? Recreate the above image using altair (10 POINTS)

In [34]:
# Recreate this image using Altair
# try to match the "538 style" as best you can (hint: look at the altair lab at the start of the semester)

# YOUR CODE HERE
from collections import Counter

seen = ["seen_{}".format(ep) for ep in episodes]
rank = ["rank_{}".format(ep) for ep in episodes]


respondents_who_seen_each_movie = sw.dropna(how="any", subset=seen)

print("People who have seen each movie : {}\n".format(respondents_who_seen_each_movie.shape[0]))

seen_all = []
for col in rank:
    rank_cnts = Counter(respondents_who_seen_each_movie[col])
    #print("{} : {}".format(col, rank_cnts["1"]/respondents_who_seen_each_movie.shape[0]))
    seen_all.append([names[col.split("_")[1]], rank_cnts["1"]/respondents_who_seen_each_movie.shape[0]])
    
seen_all_df = pd.DataFrame(seen_all, columns=["Name", "Percentage"])

seen_all_df

People who have seen each movie : 471



Name  Percentage
0     The Phantom Meanance    0.099788
1     Attack of the clones    0.038217
2      Revenge of the Sith    0.057325
3               A New Hope    0.271762
4  The Empire Strikes Back    0.358811
5   The Return of the Jedi    0.174098

In [35]:
bars = alt.Chart(seen_all_df).mark_bar(size=35).encode(
    # encode x as the percent, and hide the axis
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        # encode y using the name, use the movie name to label the axis, sort using the names_l
        'Name:N',
         axis=alt.Axis(tickCount=5, title='', labelFontSize=13,labelColor="grey", tickColor="grey", tickSize=10),
         # we give the sorting order to avoid alphabetical order
         sort=names_l
    )
)

text = bars.mark_text(
    align='left',
    baseline='middle',
    dx=3,  # Nudges text to right so it doesn't appear on top of the bar
    #color="#008fd5"
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Percentage:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
seen_all_movies = (text + bars).configure_mark(
    # we don't love the blue
    color='dodgerblue'
).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.05
).properties(
    # set the dimensions of the visualization
    width=650,
    height=280
).properties(
    # add a title
    title=alt.TitleParams("What's the Best 'Star Wars' Movie ?", subtitle=["Of 471 repondents who has seen all six film"],
                          
                         )    
    
).configure_title(anchor="start",fontSize=25,offset=20, subtitleFontSize=14)

seen_all_movies

alt.LayerChart(...)

## Make sure to *style* your visualization to match the original the best you can

We can also drill down and find out, generally, how people rate the films. Overall, fans broke into two camps: those who preferred the original three movies and those who preferred the three prequels. People who said “The Empire Strikes Back” was their favorite were also likely to rate “A New Hope” and “Return of the Jedi” higher as well. Those who rated “The Phantom Menace” as the best film were more likely to rate prequels higher.

This chart shows how often each film was rated in the top third (best or second-best), the middle third (third or fourth) or the bottom third (second-worst or worst). It’s a more nuanced take on the series:

!["Sol1"](assets/how_rate_resized.png)

** Homework note: Click [here](assets/people_rate.png) to see a version of this plot generated in Altair.

### 2.2 How people rate the 'Star Wars' movie? Recreate the above image using altair (10 POINTS)

In [36]:
# Recreate this image using altair here (10 POINTS)

# YOUR CODE HERE
names

{'EI': 'The Phantom Meanance',
 'EII': 'Attack of the clones',
 'EIII': 'Revenge of the Sith',
 'EIV': 'A New Hope',
 'EV': 'The Empire Strikes Back',
 'EVI': 'The Return of the Jedi'}

In [37]:
seen = ["seen_{}".format(ep) for ep in episodes]
rank = ["rank_{}".format(ep) for ep in episodes]

seen_all_movie = sw.dropna(how="any", subset=seen)

print("People who have seen each movie : {}\n".format(seen_all_movie.shape[0]))

top_third, middle_third, bottom_third = [],[],[]
for movie in rank:
    #print("{} : {}".format(movie, seen_all_movie[(seen_all_movie[movie]=='1') | (seen_all_movie[movie]=='2')].shape[0]/seen_all_movie.shape[0]))
    top_third.append([names[movie.split("_")[1]], seen_all_movie[(seen_all_movie[movie]=='1') | (seen_all_movie[movie]=='2')].shape[0]/seen_all_movie.shape[0]])
    middle_third.append([names[movie.split("_")[1]], seen_all_movie[(seen_all_movie[movie]=='3') | (seen_all_movie[movie]=='4')].shape[0]/seen_all_movie.shape[0]])
    bottom_third.append([names[movie.split("_")[1]], seen_all_movie[(seen_all_movie[movie]=='5') | (seen_all_movie[movie]=='6')].shape[0]/seen_all_movie.shape[0]])

People who have seen each movie : 471



In [38]:
top_third_df = pd.DataFrame(top_third, columns=["Name", "Percentage"])

top_third_df["Rating"] = "Top Third"
top_third_df["Color"] = "green"

top_third_df

Name  Percentage     Rating  Color
0     The Phantom Meanance    0.163482  Top Third  green
1     Attack of the clones    0.138004  Top Third  green
2      Revenge of the Sith    0.129512  Top Third  green
3               A New Hope    0.498938  Top Third  green
4  The Empire Strikes Back    0.641189  Top Third  green
5   The Return of the Jedi    0.428875  Top Third  green

In [39]:
middle_third_df = pd.DataFrame(middle_third, columns=["Name", "Percentage"])

middle_third_df["Rating"] = "Middle Third"
middle_third_df["Color"] = "dodgerblue"

middle_third_df

Name  Percentage        Rating       Color
0     The Phantom Meanance    0.373673  Middle Third  dodgerblue
1     Attack of the clones    0.288747  Middle Third  dodgerblue
2      Revenge of the Sith    0.401274  Middle Third  dodgerblue
3               A New Hope    0.309979  Middle Third  dodgerblue
4  The Empire Strikes Back    0.220807  Middle Third  dodgerblue
5   The Return of the Jedi    0.405520  Middle Third  dodgerblue

In [40]:
bottom_third_df = pd.DataFrame(bottom_third, columns=["Name", "Percentage"])

bottom_third_df["Rating"] = "Bottom Third"
bottom_third_df["Color"] = "red"

bottom_third_df

Name  Percentage        Rating Color
0     The Phantom Meanance    0.462845  Bottom Third   red
1     Attack of the clones    0.573248  Bottom Third   red
2      Revenge of the Sith    0.467091  Bottom Third   red
3               A New Hope    0.191083  Bottom Third   red
4  The Empire Strikes Back    0.138004  Bottom Third   red
5   The Return of the Jedi    0.165605  Bottom Third   red

In [41]:
bar1 = alt.Chart(top_third_df).mark_bar(color='green').encode(
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        'Name',
         axis=alt.Axis(tickCount=5, title='', labelFontSize=13,labelColor="grey", tickColor="grey", tickSize=10),
         sort=names_l
    ),   
    
    ).properties(
       
        width=150,
        height=270,
        title=alt.TitleParams("Top Third",fontSize=16),
        
    )

text1 = bar1.mark_text(align="left", baseline='middle', color="limegreen", dx=3).encode(
    text=alt.Text('Percentage:Q',format='.0%'),
)

bar2 = alt.Chart(middle_third_df).mark_bar(color='#008fd5').encode(
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        'Name',
         axis=None, #alt.Axis(tickCount=5, title='', labels=False, ticks=False),
         sort=names_l
    ),   
    ).properties(
        width=150,
        height=270,
        title=alt.TitleParams("Middle Third",fontSize=16)
    )

text2 = bar2.mark_text(align="left", baseline='middle',color='dodgerblue', dx=3).encode(
    text=alt.Text('Percentage:Q',format='.0%'),
)

bar3 = alt.Chart(bottom_third_df).mark_bar(color='tomato').encode(
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        'Name',
         axis=None, #alt.Axis(tickCount=5, title='', labels=False, ticks=False, ),
         sort=names_l
    ),  
    ).properties(
        width=150,
        height=270,
        title=alt.TitleParams("Bottom Third",fontSize=16)
    )

text3 = bar3.mark_text(align="left",baseline='middle',color='orangered', dx=3).encode(
    text=alt.Text('Percentage:Q',format='.0%'),
)

((bar1 + text1) | (bar2 + text2) | (bar3 + text3)).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.2
).properties(
    title= alt.TitleParams("How People Rate the 'Star Wars' Movies", subtitle=["How often each film was rated in the top, middle and bottom third", "(by 471 repondents who have seen all six films)"])    
).configure_title(fontSize=24,offset=20, subtitleFontSize=16)

alt.HConcatChart(...)

Finally, we took a boilerplate format used by political favorability polls — “Please state whether you view the following characters favorably, unfavorably, or are unfamiliar with him/her” — and asked respondents to rate characters in the series.

!["Sol1"](assets/char_ranking_resized.png)


** Homework note. Here's an example solution generated in Altair:
!["Sol1"](assets/people_rate_s.png)


### 2.3 Star Wars' Characters Favorability Ratings. Recreate the above image using altair (10 POINTS)

In [53]:
# Recreate this image using altair here (10 POINTS)


# YOUR CODE HERE
characters = ["Luke Skywalker", "Han Solo", "Princess Leia Organa", "Obi Wan Kenobi", "Yoda", "R2 D2", "C-3P0",
              "Anakin Skywalker", "Darth Vader", "Lando Calrissian", "Padme Amidala", "Boba Fett",
              "Emperor Palpatine", "Jar Jar Binks"]

reviews_at_least_one_character = sw[characters].dropna(how="all")
total = len(reviews_at_least_one_character)

print("total who have seen at least one: ", total)

favourable, neutral, unfavourable, unfamiliar = [], [], [], []
for character in characters:
    favourable.append([character, sw[sw[character].isin(["Very favorably", "Somewhat favorably"])].shape[0]/total])
    neutral.append([character, sw[sw[character].isin(["Neither favorably nor unfavorably (neutral)"])].shape[0]/total])
    unfavourable.append([character, sw[sw[character].isin(["Very unfavorably", "Somewhat unfavorably"])].shape[0]/total])
    unfamiliar.append([character, sw[sw[character].isin(["Unfamiliar (N/A)"])].shape[0]/total])
    
favourable_df = pd.DataFrame(favourable, columns=["Name", "Percentage"])
favourable_df["Type"] = "Favourable"

neutral_df = pd.DataFrame(neutral, columns=["Name", "Percentage"])
neutral_df["Type"] = "Neutral"

unfavourable_df = pd.DataFrame(unfavourable, columns=["Name", "Percentage"])
unfavourable_df["Type"] = "Unfavourable"

unfamiliar_df = pd.DataFrame(unfamiliar, columns=["Name", "Percentage"])
unfamiliar_df["Type"] = "Unfamiliar"

total who have seen at least one:  834


In [54]:
favourable_df

Name  Percentage        Type
0         Luke Skywalker    0.924460  Favourable
1               Han Solo    0.912470  Favourable
2   Princess Leia Organa    0.907674  Favourable
3         Obi Wan Kenobi    0.899281  Favourable
4                   Yoda    0.898082  Favourable
5                  R2 D2    0.895683  Favourable
6                  C-3P0    0.842926  Favourable
7       Anakin Skywalker    0.616307  Favourable
8            Darth Vader    0.576739  Favourable
9       Lando Calrissian    0.437650  Favourable
10         Padme Amidala    0.420863  Favourable
11             Boba Fett    0.348921  Favourable
12     Emperor Palpatine    0.303357  Favourable
13         Jar Jar Binks    0.290168  Favourable

In [55]:
unfavourable_df

Name  Percentage          Type
0         Luke Skywalker    0.019185  Unfavourable
1               Han Solo    0.010791  Unfavourable
2   Princess Leia Organa    0.021583  Unfavourable
3         Obi Wan Kenobi    0.017986  Unfavourable
4                   Yoda    0.019185  Unfavourable
5                  R2 D2    0.019185  Unfavourable
6                  C-3P0    0.035971  Unfavourable
7       Anakin Skywalker    0.146283  Unfavourable
8            Darth Vader    0.300959  Unfavourable
9       Lando Calrissian    0.085132  Unfavourable
10         Padme Amidala    0.110312  Unfavourable
11             Boba Fett    0.169065  Unfavourable
12     Emperor Palpatine    0.230216  Unfavourable
13         Jar Jar Binks    0.366906  Unfavourable

In [56]:
neutral_df

Name  Percentage     Type
0         Luke Skywalker    0.045564  Neutral
1               Han Solo    0.052758  Neutral
2   Princess Leia Organa    0.057554  Neutral
3         Obi Wan Kenobi    0.051559  Neutral
4                   Yoda    0.061151  Neutral
5                  R2 D2    0.068345  Neutral
6                  C-3P0    0.094724  Neutral
7       Anakin Skywalker    0.161871  Neutral
8            Darth Vader    0.100719  Neutral
9       Lando Calrissian    0.282974  Neutral
10         Padme Amidala    0.248201  Neutral
11             Boba Fett    0.297362  Neutral
12     Emperor Palpatine    0.255396  Neutral
13         Jar Jar Binks    0.196643  Neutral

In [57]:
unfamiliar_df

Name  Percentage        Type
0         Luke Skywalker    0.007194  Unfamiliar
1               Han Solo    0.017986  Unfamiliar
2   Princess Leia Organa    0.009592  Unfamiliar
3         Obi Wan Kenobi    0.020384  Unfamiliar
4                   Yoda    0.011990  Unfamiliar
5                  R2 D2    0.011990  Unfamiliar
6                  C-3P0    0.017986  Unfamiliar
7       Anakin Skywalker    0.062350  Unfamiliar
8            Darth Vader    0.011990  Unfamiliar
9       Lando Calrissian    0.177458  Unfamiliar
10         Padme Amidala    0.196643  Unfamiliar
11             Boba Fett    0.158273  Unfamiliar
12     Emperor Palpatine    0.187050  Unfamiliar
13         Jar Jar Binks    0.130695  Unfamiliar

In [47]:
bar1 = alt.Chart(favourable_df).mark_bar(color='green').encode(
    x=alt.X('Percentage', axis=None),
    y=alt.Y('Name', axis=alt.Axis(tickCount=5, title='', labelFontSize=13,labelColor="grey", tickColor="grey", tickSize=10),
    sort=characters
    ),   
    
    ).properties(
       
        width=120,
        height=300,
        title=alt.TitleParams("Favourable", fontSize=18),
        
    )

text1 = bar1.mark_text(align="left", baseline='middle', color="limegreen", dx=3).encode(
    text=alt.Text('Percentage:Q',format='.0%'),
)

bar2 = alt.Chart(neutral_df).mark_bar(color='#008fd5').encode(
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        'Name',
         axis=None, #alt.Axis(tickCount=5, title='', labels=False, ticks=False),
         sort=characters
        
    ),   
    ).properties(
        width=120,
        height=300,
        title=alt.TitleParams("Neutral", fontSize=18)
    )

text2 = bar2.mark_text(align="left", baseline='middle',color='dodgerblue', dx=3).encode(
    text=alt.Text('Percentage:Q',format='.0%'),
)

bar3 = alt.Chart(unfavourable_df).mark_bar(color='tomato').encode(
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        'Name',
         axis=None, #alt.Axis(tickCount=5, title='', labels=False, ticks=False, ),
        sort=characters
         
    ),  
    
    ).properties(
        width=120,
        height=300,
        title=alt.TitleParams("Unfavourable", fontSize=18)
    )

text3 = bar3.mark_text(align="left",baseline='middle',color='orangered', dx=3).encode(
    text=alt.Text('Percentage:Q',format='.0%'),
)

bar4 = alt.Chart(unfamiliar_df).mark_bar(color='grey').encode(
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        'Name',
         axis=None, #alt.Axis(tickCount=5, title='', labels=False, ticks=False, ),
         sort=characters
    ),  
    
    ).properties(
        width=120,
        height=300,
        title=alt.TitleParams("Unfamiliar", fontSize=18)
    )

text4 = bar4.mark_text(align="left",baseline='middle',color='grey', dx=3).encode(
    text=alt.Text('Percentage:Q',format='.0%'),
)

((bar1 + text1) | (bar2 + text2) | (bar3 + text3) | (bar4 + text4)).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.2
).properties(
    title=alt.TitleParams("'Star Wars' Characters Favourability Ratings", subtitle=["By %d respondents"%(total)])    
).configure_title(fontSize=24,offset=20, subtitleFontSize=16)

alt.HConcatChart(...)

You read that correctly. Jar Jar Binks has a lower favorability rating than the actual personification of evil in the galaxy.

And for those of you who want to know the impact that [historical revisionism](http://en.wikipedia.org/wiki/Han_shot_first) can have on a society:

!["Sol1"](assets/shot_first_article_resized.png)


** Homework note: Click [here](assets/shot_first.png) to see a version of this plot generated in Altair. You may find that you don't get 834 rows (as 538 did) but the percents should still work.

### 2.4 Who shot first? Recreate the above image using altair (10 POINTS)

In [48]:
# Recreate this image using altair here (10 POINTS)


# YOUR CODE HERE
reviews_at_least_one_character = sw.dropna(how="all", subset=rank)
total = len(reviews_at_least_one_character)
cnts = Counter(reviews_at_least_one_character["Which character shot first?"].values)
total = total - cnts[np.nan]
print("Total who have seen at least one: ", total,"\n")

who_shot_first = []
for character, cnt in cnts.items():
    #print("{} : {}".format(character, cnt/826))
    if isinstance(character, str):
        who_shot_first.append([character, cnt/total])
    
who_shot_first_df = pd.DataFrame(who_shot_first, columns=["Name", "Percentage"]).set_index("Name").loc[["Han", "Greedo", "I don't understand this question"]].reset_index()

who_shot_first_df

Total who have seen at least one:  828 



Name  Percentage
0                               Han    0.392512
1                            Greedo    0.237923
2  I don't understand this question    0.369565

In [49]:
bars = alt.Chart(who_shot_first_df).mark_bar(size=35).encode(
    # encode x as the percent, and hide the axis
    x=alt.X(
        'Percentage',
        axis=None),
    y=alt.Y(
        # encode y using the name, use the movie name to label the axis, sort using the names_l
        'Name:N',
         axis=alt.Axis(tickCount=5, title='', labelFontSize=13,labelColor="grey", tickColor="grey", tickSize=10),
         # we give the sorting order to avoid alphabetical order
         sort=names_l
    )
)

text = bars.mark_text(
    align='left',
    baseline='middle',
    dx=3,  # Nudges text to right so it doesn't appear on top of the bar
    #color="#008fd5"
).encode(
    # we'll use the percentage as the text
    text=alt.Text('Percentage:Q',format='.0%')
)

# finally, we're going to combine the bars and the text and do some styling
seen_all_movies = (text + bars).configure_mark(
    # we don't love the blue
    color='dodgerblue'
).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.1
).properties(
    # set the dimensions of the visualization
    width=700,
    height=150
).properties(
    # add a title
    title=alt.TitleParams("Who Shot First ?", subtitle=["According to %d repondents"%(total)],
                          
                         )    
    
).configure_title(anchor="start", fontSize=25,offset=20, subtitleFontSize=14)

seen_all_movies

alt.LayerChart(...)

### 2.5.1 Make your own (15 points/ 10 points plot + 5 justification)

Propose and code an alternative visualization for one of the visualizations *already in the article*. Add a short paragraph describing why your visualization is more (or less) *effective* based on principles of perception/cognition. 

If you feel your visualization is worse, that's ok! Just tell us why.

In [132]:
combined_df = pd.concat((top_third_df, middle_third_df, bottom_third_df))

combined_df

Name  Percentage        Rating       Color
0     The Phantom Meanance    0.163482     Top Third       green
1     Attack of the clones    0.138004     Top Third       green
2      Revenge of the Sith    0.129512     Top Third       green
3               A New Hope    0.498938     Top Third       green
4  The Empire Strikes Back    0.641189     Top Third       green
5   The Return of the Jedi    0.428875     Top Third       green
0     The Phantom Meanance    0.373673  Middle Third  dodgerblue
1     Attack of the clones    0.288747  Middle Third  dodgerblue
2      Revenge of the Sith    0.401274  Middle Third  dodgerblue
3               A New Hope    0.309979  Middle Third  dodgerblue
4  The Empire Strikes Back    0.220807  Middle Third  dodgerblue
5   The Return of the Jedi    0.405520  Middle Third  dodgerblue
0     The Phantom Meanance    0.462845  Bottom Third         red
1     Attack of the clones    0.573248  Bottom Third         red
2      Revenge of the Sith    0.467091  Bottom Third         red
3               A New Hope    0.191083  Bottom Third         red
4  The Empire Strikes Back    0.138004  Bottom Third         red
5   The Return of the Jedi    0.165605  Bottom Third         red

In [135]:
bars = alt.Chart(combined_df).mark_bar().encode(
    x=alt.X('Percentage:Q', stack='zero', axis=None),
    y=alt.Y('Name:O', sort=names_l, title="", axis=alt.Axis(tickCount=5, title='', labelFontSize=13,labelColor="grey", tickColor="grey", tickSize=10),),
    color=alt.Color('Rating',
        #sort = alt.EncodingSortField( 'order', order = 'ascending' ),
        scale = alt.Scale(domain=['Top Third', 'Middle Third', 'Bottom Third'], range=['limegreen', 'dodgerblue', 'tomato']),
        legend = alt.Legend(title="Rating")
    ),
)

text = alt.Chart(combined_df).mark_text(dx=-25, dy=3, color='black').encode(
    x=alt.X('Percentage:Q', stack='zero', axis=None),
    y=alt.Y('Name:O', sort=names_l),
    text=alt.Text('Percentage:Q', format='.1%')
)

(bars + text).configure_view(
    # we don't want a stroke around the bars
    strokeWidth=0
).configure_scale(
    # add some padding
    bandPaddingInner=0.2
).properties(
    width=670,
    height=400,
    title= alt.TitleParams("How People Rate the 'Star Wars' Movies", subtitle=["How often each film was rated in the top, middle and bottom third", "(by 471 repondents who have seen all six films)"])
).configure_title(anchor="start", fontSize=24,offset=20, subtitleFontSize=16)

alt.LayerChart(...)

#### Justification

Above we have recreated a visualization from section 2.2 (How People Rate the 'Star Wars' Movies). The reason behind creating the rating distribution this way was to combine ratings to effectively compare them. Above chart lays rating groups next to each other which can let us relatively compare the distribution of rating groups for each movie and it also let us very easily compare rating group distributions across different movies.

Above visualization clearly help us understand that 3 movies (A New Hope, The Empire Strikes Back and The return of the Jedi) were more top rated. We can easily notice that The Empire Strikes Back was liked by many followed by A New Hope and The Return of the Jedi.

### 2.5.2 Make your own (15 points/ 10 points plot + 5 justification)
Propose and code a *new visualization* to complement a part of the article. Add a short paragraph justifying your decisions in terms of Perception/Cognition processes.

If you feel your visualization is worse, that's ok! Just tell us why.

In [ ]:
# YOUR CODE HERE
seen_single_movie = sw.dropna(how="all", subset=seen)

data = []
for movie in seen:
    intermediate_df = seen_single_movie[~seen_single_movie[movie].isnull()]
    #expanded_cnt = Counter(intermediate_df["Are you familiar with the Expanded Universe?"])
    age_cnt = Counter(intermediate_df["Age"])
    
    #data.append([names[movie.split("_")[1]], "Expanded_Yes", expanded_cnt["Yes"]])
    #data.append([names[movie.split("_")[1]], "Expanded_No", expanded_cnt["No"]])
    data.append([names[movie.split("_")[1]], "18-29", age_cnt["18-29"]])
    data.append([names[movie.split("_")[1]], "30-44", age_cnt["30-44"]])
    data.append([names[movie.split("_")[1]], "45-60", age_cnt["45-60"]])
    data.append([names[movie.split("_")[1]], "> 60", age_cnt["> 60"]])
    
combined_df = pd.DataFrame(data, columns=["Movie", "Age Group", "Value"])

combined_df

In [176]:
# YOUR CODE HERE
seen_single_movie = sw.dropna(how="all", subset=seen)

data = []
for movie in seen:
    intermediate_df = seen_single_movie[~seen_single_movie[movie].isnull()]
    #expanded_cnt = Counter(intermediate_df["Are you familiar with the Expanded Universe?"])
    education_cnt = Counter(intermediate_df["Education"])
    
    #data.append([names[movie.split("_")[1]], "Expanded_Yes", expanded_cnt["Yes"]])
    #data.append([names[movie.split("_")[1]], "Expanded_No", expanded_cnt["No"]])
    data.append([names[movie.split("_")[1]], "Less than high school degree", education_cnt["Less than high school degree"]])
    data.append([names[movie.split("_")[1]], "High school degree", education_cnt["High school degree"]])
    data.append([names[movie.split("_")[1]], "Some college or Associate degree", education_cnt["Some college or Associate degree"]])
    data.append([names[movie.split("_")[1]], "Bachelor degree", education_cnt["Bachelor degree"]])
    data.append([names[movie.split("_")[1]], "Graduate degree", education_cnt["Graduate degree"]])
    
combined_df = pd.DataFrame(data, columns=["Movie", "Education", "Value"])

combined_df

Movie                         Education  Value
0      The Phantom Meanance      Less than high school degree      3
1      The Phantom Meanance                High school degree     57
2      The Phantom Meanance  Some college or Associate degree    211
3      The Phantom Meanance                   Bachelor degree    206
4      The Phantom Meanance                   Graduate degree    179
5      Attack of the clones      Less than high school degree      3
6      Attack of the clones                High school degree     48
7      Attack of the clones  Some college or Associate degree    186
8      Attack of the clones                   Bachelor degree    170
9      Attack of the clones                   Graduate degree    149
10      Revenge of the Sith      Less than high school degree      3
11      Revenge of the Sith                High school degree     48
12      Revenge of the Sith  Some college or Associate degree    183
13      Revenge of the Sith                   Bachelor degree    163
14      Revenge of the Sith                   Graduate degree    139
15               A New Hope      Less than high school degree      3
16               A New Hope                High school degree     53
17               A New Hope  Some college or Associate degree    180
18               A New Hope                   Bachelor degree    195
19               A New Hope                   Graduate degree    163
20  The Empire Strikes Back      Less than high school degree      3
21  The Empire Strikes Back                High school degree     61
22  The Empire Strikes Back  Some college or Associate degree    227
23  The Empire Strikes Back                   Bachelor degree    243
24  The Empire Strikes Back                   Graduate degree    207
25   The Return of the Jedi      Less than high school degree      3
26   The Return of the Jedi                High school degree     60
27   The Return of the Jedi  Some college or Associate degree    223
28   The Return of the Jedi                   Bachelor degree    234
29   The Return of the Jedi                   Graduate degree    201

In [177]:
alt.Chart(combined_df).mark_bar().encode(
    x=alt.X('Education:O', title="Education"),
    y=alt.Y('Value:Q', title="Count"),
    color=alt.Color('Education:N',
        #scale = alt.Scale(domain=['18-29', '30-44', '45-60','> 60'], range=['limegreen', 'dodgerblue', 'tomato', 'grey']),
        legend = alt.Legend(title="Rating")
    ),
    column='Movie:N'
).properties(
    width=115,
    height=400,
    title=alt.TitleParams("Education Distribution of People per Movie", fontSize=20)
)

alt.Chart(...)

#### Justification

* A New Hope - 1977 - Ep. 4
* Empire Strikes Back - 1980 - Ep. 5
* The Return of Jedi - 1983 - Ep. 6
* The Phantom Menace - 1999 - Ep. 1
* Attack of the Clones - 2002 - Ep. 2
* Revenge of Sith - 2005 - Ep. 3

We have created a side by side bar chart showing distribution of education of people who liked different episode of the movies. We have created this chart to see whether there is any pattern related to education of people who like star wars movies. We wanted to know which education group watches star wars movies more. We can notice from the above chart that majority of star wars watcher has bachelor or graduate degree. Very few people have less than high school, some college and high school degrees. This can also be coincidence because people with bachelor or graduate degree will likely have access to computer and given answer of survey if survey was conducted online. We can also notice from the chart that The Empire Strikes Back, The Phantom Meanance, and The Returns of the Jedi are most watched movies.